misc/peer-id-generator/src/main.rs

```diff
@@ -0,0 +1,81 @@
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+extern crate libp2p_core;
+extern crate libp2p_secio;
+extern crate num_cpus;
+extern crate rand;
+
+use libp2p_core::PeerId;
+use libp2p_secio::SecioKeyPair;
+use std::{env, str, thread, time::Duration};
+
+fn main() {
+    let prefix = match env::args().nth(1) {
+        Some(prefix) => prefix,
+        None => {
+            println!(
+                "Usage: {} <prefix>",
+                env::current_exe().unwrap().file_name().unwrap().to_str().unwrap()
+            );
+            return;
+        }
+    };
+
+    // The base58 alphabet is not necessarily obvious.
+    const ALPHABET: &'static [u8] = b"123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
+    if prefix.as_bytes().iter().any(|c| !ALPHABET.contains(c)) {
+        println!("Prefix {} is not valid base58", prefix);
+        return;
+    }
+
+    // Due to the fact that a peer id uses a SHA-256 multihash, it always starts with the
+    // bytes 0x1220, meaning that only some characters are valid.
+    if !prefix.is_empty() {
+        const ALLOWED_FIRST_BYTE: &'static [u8] = b"NPQRSTUVWXYZ";
+        if !ALLOWED_FIRST_BYTE.contains(&prefix.as_bytes()[0]) {
+            println!("Prefix {} is not reachable", prefix);
+            println!(
+                "Only the following bytes are possible as first byte: {}",
+                str::from_utf8(ALLOWED_FIRST_BYTE).unwrap()
+            );
+            return;
+        }
+    }
+
+    // Find peer IDs in a multithreaded fashion.
+    for _ in 0..num_cpus::get() {
+        let prefix = prefix.clone();
+        thread::spawn(move || loop {
+            let private_key: [u8; 32] = rand::random();
+            let generated = SecioKeyPair::secp256k1_raw_key(private_key).unwrap();
+            let peer_id: PeerId = generated.to_public_key().into_peer_id();
+            let base58 = peer_id.to_base58();
+            if base58[2..].starts_with(&prefix) {
+                println!("Found {:?}", peer_id);
+                println!("=> Private key = {:?}", private_key);
+            }
+        });
+    }
+
+    loop {
+        thread::sleep(Duration::from_secs(3600));
```

Comment:
Instead of executing for a certain amount of time, would you not rather want to generate a specific number of keys? How about a second command line parameter to specific that number and then maybe utilise `rayon` with `    (0 .. number).into_par_iter().for_each(move |_| ...)`?

Reply:
Generating keys is super slow, so I don't think it really makes sense to stop after a certain number.
Do you think I should use `rayon`?

Reply:
Rayon works well but I was only suggesting it in the context of "I want *n* peer IDs". If you rather prefer time-based execution the existing approach is fine I think.